go-controller/pkg/ovn/obj_retry.go

```diff
@@ -1218,7 +1464,11 @@ func (oc *Controller) WatchResource(objectsToRetry *retryObjs) (*factory.Handler
 						klog.Errorf("Failed to update resource %v, old=%s, new=%s, error: %v",
 							objectsToRetry.oType, oldKey, newKey, err)
 						oc.recordErrorEvent(objectsToRetry.oType, newer, err)
-						objectsToRetry.initRetryObjWithAdd(newer, newKey)
+						if resourceNeedsUpdate(objectsToRetry.oType) {
```

Comment:
Aha that's true... 

We were trying to avoid a delete and replace it with an update, but the delete will happen anyway in step 1a:

```
				retryEntry := objectsToRetry.getObjRetryEntry(oldKey)
				if retryEntry != nil && retryEntry.oldObj != nil {
					// [step 1a] there is a retry entry marked for deletion
					klog.Infof("Found old retry object for %v %s: will delete it",
						objectsToRetry.oType, oldKey)
					if err := oc.deleteResource(objectsToRetry, retryEntry.oldObj,
						retryEntry.config); err != nil {
						klog.Errorf("Failed to delete stale object %s, during update: %v", oldKey, err)
						oc.recordErrorEvent(objectsToRetry.oType, retryEntry.oldObj, err)
						objectsToRetry.initRetryObjWithAdd(newer, newKey)
						objectsToRetry.unSkipRetryObj(oldKey)
						return
					}
					// remove the old object from retry entry since it was correctly deleted
					objectsToRetry.removeDeleteFromRetryObj(oldKey)
```
So if `resourceNeedsUpdate` is true, we might want to skip this deletion...

Reply:
yeah

Reply:
I messed up the flags above, let me correct them :D

Reply:
Now with the latest change we skip delete if it needsUpdate
```
				if retryEntry != nil && retryEntry.oldObj != nil &&
					!resourceNeedsUpdate(objectsToRetry.oType) {
```

However, consider the opposite scenario. Let's assume A and B are different instances of the same keyed object, and v1, v2, etc are revisions of each instance. So lets say this happens:

1. object A exists
2. object A is deleted, but it fails added to retry delete retry->old == A, retry->new=nil
3. object Bv1 is created, but it fails to delete A, add Bv1 to retry add, retry->old == A, retry->new==Bv1
4. object Bv1 is updated to Bv2, now because of this logic the A is no longer deleted, and lets assume update fails, now retry->old== Bv1, retry new==Bv2

Consequently A may never be deleted. Is this OK for the controllers that return true for resourceNeedsUpdate?

Reply:
solution for now is to store the "update old object" in retryObject->config, and store the object that we wanted to deleted still in old. It's not perfect and we will fix it later, but solves the above issue for now.